README.md
# remark-blockquotes

Add custom blockquotes to your markdown. This package is intended to work over remark using mdast \
It is based on the GFM implementation: https://github.com/orgs/community/discussions/16925

## Installation

```bash
npm i remark-blockquotes
```

## Usage

### Block Types

#### Note block

This block is intended to be used for notes, tips, etc. If the block type specified does not exist, it will be rendered as a note block.

```markdown
> **Note** \
> Lorem ipsum
```

Will be transformed to:

```html
<blockquote class="blockquote blockquote--note" data-type="note" data-original-title="Note">
  <p>
    <span class="blockquote__header">
      <svg class="octicon octicon-info" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" width="16" height="16"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>
      <strong>Note</strong>
    </span> <br>
    Lorem ipsum
  </p>
</blockquote>
```

#### Warning block

This block is intended to be used for warnings, alerts, etc.

```markdown
> **Warning** \
> Lorem ipsum
```

Will be transformed to:

```html
<blockquote class="blockquote blockquote--warning" data-type="warning" data-original-title="Warning">
  <p>
    <span class="blockquote__header">
      <svg class="octicon octicon-info" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" width="16" height="16"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>
      <strong>Warning</strong>
    </span> <br>
    Lorem ipsum
  </p>
</blockquote>
```

#### Error block

This block is intended to be used for errors, failures, etc.

```markdown
> **Error**
> Lorem ipsum \
> Dolor sit amet
```

Will be transformed to:

```html
<blockquote class="blockquote blockquote--error" data-type="error" data-original-title="Error">
  <p>
    <span class="blockquote__header">
      <svg class="octicon octicon-circle-slash" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" width="16" height="16"><path d="M8 0a8 8 0 1 1 0 16A8 8 0 0 1 8 0ZM3.965 13.096a6.5 6.5 0 0 0 9.131-9.131ZM1.5 8a6.474 6.474 0 0 0 1.404 4.035l9.131-9.131A6.499 6.499 0 0 0 1.5 8Z"></path></svg>
      <strong>Error</strong>
    </span> <br>
    Lorem ipsum <br>
    Dolor sit amet
  </p>
</blockquote>
```


### Custom Header Message

You can also customize the header message by adding a custom header message after the block type between brackets.

```markdown
> **Error[Ay mecachis]** \
> Lorem ipsum \
> Dolor sit amet
```

### Adding the script

And our script looks as follows:

```javascript
const remark = require('remark');

remark()
  .use(require('remark-blockquotes'))
  .use(require('remark-html'))
  .process(src, (err, file) => console.log(String(file)));
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

## 1.0.0 (2023-07-12)


.eslintrc.js
module.exports = {
  root: true,
  parserOptions: {
    ecmaVersion: 12,
    sourceType: 'module',
    ecmaFeatures: {
      modules: true,
      jsx: true,
    },
  },
  env: {
    es2021: true,
    browser: true,
    node: true,
    commonjs: true,
  },
  extends: ['eslint:recommended', 'airbnb', 'plugin:prettier/recommended'],
  rules: {
    'no-console': ['error', { allow: ['warn', 'error', 'info'] }],
    'prefer-const': ['error', {}],
    'no-plusplus': ['error', { allowForLoopAfterthoughts: true }],
    'jsx-a11y/no-static-element-interactions': [0],
    'jsx-a11y/click-events-have-key-events': [0],
    'jsx-a11y/media-has-caption': [0],
    'class-methods-use-this': ['warn', {}],
    'no-restricted-exports': ['error', { restrictedNamedExports: ['then'] }],
    'no-underscore-dangle': [
      'warn',
      {
        allowAfterThis: true,
        allowAfterSuper: true,
      },
    ],
    'comma-dangle': ['error', 'always-multiline'],
  },
  overrides: [
    {
      files: ['.settings/**/*'],
      rules: {
        'import/no-extraneous-dependencies': [
          'error',
          {
            devDependencies: true,
            optionalDependencies: false,
            peerDependencies: true,
            packageDir: ['./'],
          },
        ],
      },
    },
  ],
  ignorePatterns: ['.parcel-cache', 'node_modules', '!.settings'],
};


src/index.js
// <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" width="16" height="16"><path d="M8 0a8 8 0 1 1 0 16A8 8 0 0 1 8 0ZM3.965 13.096a6.5 6.5 0 0 0 9.131-9.131ZM1.5 8a6.474 6.474 0 0 0 1.404 4.035l9.131-9.131A6.499 6.499 0 0 0 1.5 8Z"></path></svg>

// from github.com/syntax-tree/unist-util-map/blob/bb0567f651517b2d521af711d7376475b3d8446a/index.js
const Octicons = require('./octicons');

const map = (tree, iteratee) => {
  const preorder = (node, index, parent) => {
    const newNode = iteratee(node, index, parent);

    if (Array.isArray(newNode.children)) {
      newNode.children = newNode.children.map((child, idxChild) => {
        return preorder(child, idxChild, node);
      });
    }

    return newNode;
  };

  return preorder(tree, null, null);
};

const toKebabCase = (str) =>
  str &&
  str
    .match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
    .map((x) => x.toLowerCase())
    .join('-');

const parseQuoteHeader = (text) => {
  const REGEX = /([\w\s]+)(?:\[(.*)])?/g;
  const [, type, message] = REGEX.exec(text);

  const title = {
    type: 'strong',
    children: [
      {
        type: 'text',
        value: message || type,
      },
    ],
  };

  let svgBlock;

  let quoteType = toKebabCase(type);

  switch (quoteType) {
    case 'error':
      svgBlock = Octicons.CircleSlash;
      break;
    case 'warning':
      svgBlock = Octicons.Alert;
      break;
    case 'note':
      svgBlock = Octicons.Info;
      break;
    default:
      svgBlock = Octicons.Info;
      quoteType = 'note';
      break;
  }

  const quoteHeader = {
    type: quoteType,
    blocks: [
      {
        type: 'parent',
        children: [svgBlock, title],
        data: {
          hName: 'span',
          hProperties: {
            class: 'blockquote__header',
          },
        },
      },
    ],
    attributes: {
      class: `blockquote blockquote--${quoteType}`,
      'data-type': quoteType,
      'data-original-title': type,
    },
  };
  return quoteHeader;
};

module.exports = () => (tree) => {
  return map(tree, (node) => {
    const { children = [] } = node;
    if (node.type !== 'blockquote') {
      return node;
    }

    const [block] = children;
    if (block.type !== 'paragraph') {
      return node;
    }

    const [first] = block.children;
    let quoteHeader = null;
    if (first.type === 'strong' && first.children.length > 0 && first.children[0].type === 'text') {
      quoteHeader = parseQuoteHeader(first.children[0].value);
      block.children = [...quoteHeader.blocks, ...block.children.slice(1)];

      // eslint-disable-next-line no-param-reassign
      node.data = {
        ...node.data,
        hProperties: {
          ...quoteHeader.attributes,
        },
      };
    }

    return node;
  });
};


src/octicons.js
const CircleSlash = {
  type: 'parent',
  children: [
    {
      type: 'parent',
      data: {
        hName: 'path',
        hProperties: {
          d: 'M8 0a8 8 0 1 1 0 16A8 8 0 0 1 8 0ZM3.965 13.096a6.5 6.5 0 0 0 9.131-9.131ZM1.5 8a6.474 6.474 0 0 0 1.404 4.035l9.131-9.131A6.499 6.499 0 0 0 1.5 8Z',
        },
      },
    },
  ],
  data: {
    hName: 'svg',
    hProperties: {
      class: 'octicon octicon-circle-slash',
      xmlns: 'http://www.w3.org/2000/svg',
      viewBox: '0 0 16 16',
      width: '16',
      height: '16',
    },
  },
};

const Info = {
  type: 'parent',
  children: [
    {
      type: 'parent',
      data: {
        hName: 'path',
        hProperties: {
          d: 'M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z',
        },
      },
    },
  ],
  data: {
    hName: 'svg',
    hProperties: {
      class: 'octicon octicon-info',
      xmlns: 'http://www.w3.org/2000/svg',
      viewBox: '0 0 16 16',
      width: '16',
      height: '16',
    },
  },
};

const Alert = {
  type: 'parent',
  children: [
    {
      type: 'parent',
      data: {
        hName: 'path',
        hProperties: {
          d: 'M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z',
        },
      },
    },
  ],
  data: {
    hName: 'svg',
    hProperties: {
      class: 'octicon octicon-info',
      xmlns: 'http://www.w3.org/2000/svg',
      viewBox: '0 0 16 16',
      width: '16',
      height: '16',
    },
  },
};

module.exports = {
  CircleSlash,
  Info,
  Alert,
};
